File: algo/19. Remove Nth Node From End of List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    if (!head) {
        return head;
    }
    
    var prevHead = new ListNode(0);
    prevHead.next = head;
    var currNode = head;
    var listLength = 1;
    while (currNode.next) {
        listLength ++;
        currNode = currNode.next;
    }
    
    var targetIndex = listLength - n;
    currNode = prevHead;
    while(targetIndex>0) {
        currNode = currNode.next;
        targetIndex--;
    }
    
    currNode.next = currNode.next.next;
    
    return prevHead.next;
        
};

File: round3/19. Remove Nth Node From End of List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode RemoveNthFromEnd(ListNode head, int n) {
        ListNode left = head, right = head;
        for (int i = 0; i < n; i ++) {
            right = right.next;
        }
        
        if (right == null) return head.next;
        
        while (right.next != null) {
            right = right.next;
            left = left.next;
        }
        
        left.next = left.next.next;
        return head;
    }
}